tests/unit/waveform/test_digital_waveform.py

```diff
@@ -0,0 +1,1285 @@
+    assert_type(waveform.timing.start_time, Union[bt.DateTime, dt.datetime, ht.datetime])
+    assert_type(waveform.timing.time_offset, Union[bt.TimeDelta, dt.timedelta, ht.timedelta])
+    assert waveform.timing.sample_interval == sample_interval
+    assert waveform.timing.timestamp == timestamp
+    assert waveform.timing.start_time == timestamp + time_offset
+    assert waveform.timing.time_offset == time_offset
+
+
+def test___hightime___waveform_with_timing___static_type_erased() -> None:
+    sample_interval = ht.timedelta(milliseconds=1)
+    timestamp = ht.datetime.now(dt.timezone.utc)
+    time_offset = ht.timedelta(microseconds=1)
+    waveform = DigitalWaveform(
+        timing=Timing.create_with_regular_interval(sample_interval, timestamp, time_offset)
+    )
+
+    assert_type(waveform.timing.sample_interval, Union[bt.TimeDelta, dt.timedelta, ht.timedelta])
+    assert_type(waveform.timing.timestamp, Union[bt.DateTime, dt.datetime, ht.datetime])
+    assert_type(waveform.timing.start_time, Union[bt.DateTime, dt.datetime, ht.datetime])
+    assert_type(waveform.timing.time_offset, Union[bt.TimeDelta, dt.timedelta, ht.timedelta])
+    assert waveform.timing.sample_interval == sample_interval
+    assert waveform.timing.timestamp == timestamp
+    assert waveform.timing.start_time == timestamp + time_offset
+    assert waveform.timing.time_offset == time_offset
+
+
+@pytest.mark.parametrize(
+    "timing",
+    [
+        Timing.create_with_regular_interval(
+            bt.TimeDelta(1e-3), bt.DateTime(2025, 1, 1, tzinfo=dt.timezone.utc), bt.TimeDelta(1e-6)
+        ),
+        Timing.create_with_regular_interval(
+            dt.timedelta(milliseconds=1),
+            dt.datetime(2025, 1, 1, tzinfo=dt.timezone.utc),
+            dt.timedelta(microseconds=1),
+        ),
+        Timing.create_with_regular_interval(
+            ht.timedelta(milliseconds=1),
+            ht.datetime(2025, 1, 1, tzinfo=dt.timezone.utc),
+            dt.timedelta(microseconds=1),
+        ),
+    ],
+)
+def test___polymorphic_timing___get_timing_properties___behaves_polymorphically(
+    timing: Timing[Any, Any, Any],
+) -> None:
+    waveform = DigitalWaveform(timing=timing)
+
+    assert waveform.timing.sample_interval.total_seconds() == pytest.approx(1e-3)
+    assert (
+        waveform.timing.timestamp.year,
+        waveform.timing.timestamp.month,
+        waveform.timing.timestamp.day,
+    ) == (2025, 1, 1)
+    assert waveform.timing.time_offset.total_seconds() == pytest.approx(1e-6)
+
+
+###############################################################################
+# append array
+###############################################################################
+def test___empty_ndarray___append___no_effect() -> None:
+    waveform = DigitalWaveform.from_lines([[0], [1], [2]], np.uint8)
+    array = np.array([], np.uint8)
+
+    waveform.append(array)
+
+    assert waveform.data.tolist() == [[0], [1], [2]]
```

Comment:
Maybe there should be a `data_1d` property that returns a 1D view (or throws if signal_count > 1)

Reply:
I'm going to apply "There should be one-- and preferably only one --obvious way to do it." as "you should call wfm.data.reshape(...)" for now.